File: sce-forge-runtime/cpp/tests/conformance/numerical_conformance_test.cpp
// SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2025 newmassrael
//
// Cross-language numerical conformance harness (C++ half).
//
// The fixture headers included below are produced at CMake configure time by
// the generate_forge_fixtures custom target, which invokes sce-codegen on
// the SCXML files under tests/forge/resources/ and writes the C++ output
// into ${CMAKE_CURRENT_BINARY_DIR}/generated. That directory is added to
// this translation unit's include path by CMakeLists.txt. No committed C++
// goldens are consumed — the single source of truth is the SCXML and the
// codegen.
//
// The generated fixtures are exercised against the reference vectors in
// tests/forge/conformance/numerical_reference.json (path provided via the
// REFERENCE_JSON_PATH compile-time define), the same file used by the Rust,
// Python, Kotlin, and Go conformance tests.

#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <string>
#include <vector>

#include <nlohmann/json.hpp>

#include "condition_programming.h"
#include "condition_range.h"
#include "condition_threshold.h"
#include "filter_debounce.h"
#include "filter_moving_average.h"
#include "interpolation_1d_linear.h"
#include "interpolation_2d_bilinear.h"
#include "observer_coolant.h"
#include "procedure_diamond.h"
#include "procedure_linear.h"
#include "procedure_startup_check.h"
#include "transform_bitwise.h"
#include "transform_multi_output.h"
#include "transform_temperature.h"

using json = nlohmann::json;

namespace {

int g_failures = 0;

#define CHECK(cond, ...)                               \
    do {                                               \
        if (!(cond)) {                                 \
            std::printf("FAIL: ");                     \
            std::printf(__VA_ARGS__);                  \
            std::printf("\n");                         \
            ++g_failures;                              \
        }                                              \
    } while (0)

void assert_close(double actual, double expected, double tol, const char *label) {
    double diff = std::fabs(actual - expected);
    CHECK(diff <= tol,
          "%s: actual=%.17g expected=%.17g diff=%.17g tol=%.17g",
          label,
          actual,
          expected,
          diff,
          tol);
}

json load_reference() {
    std::ifstream in(REFERENCE_JSON_PATH);
    if (!in.is_open()) {
        std::fprintf(stderr,
                     "cannot open reference JSON at %s\n",
                     REFERENCE_JSON_PATH);
        std::exit(2);
    }
    json ref;
    in >> ref;
    return ref;
}

void test_interpolation_1d_linear(const json &ref, double tol) {
    const auto &spec = ref["pure_functions"]["interpolation_1d_linear"];
    for (const auto &c : spec["cases"]) {
        auto rpm = c["args"][0].get<std::uint16_t>();
        auto expected = c["expected"].get<double>();
        double actual =
            SCE::Generated::Interpolation1dLinear::Interpolation1dLinear::lookup(rpm);
        char label[128];
        std::snprintf(label, sizeof(label), "interpolation_1d_linear(%u)",
                      static_cast<unsigned>(rpm));
        assert_close(actual, expected, tol, label);
    }
}

void test_interpolation_2d_bilinear(const json &ref, double tol) {
    const auto &spec = ref["pure_functions"]["interpolation_2d_bilinear"];
    for (const auto &c : spec["cases"]) {
        auto rpm = c["args"][0].get<std::uint16_t>();
        auto load = c["args"][1].get<std::uint8_t>();
        auto expected = c["expected"].get<double>();
        double actual =
            SCE::Generated::Interpolation2dBilinear::Interpolation2dBilinear::lookup(
                rpm, load);
        char label[128];
        std::snprintf(label, sizeof(label),
                      "interpolation_2d_bilinear(%u, %u)",
                      static_cast<unsigned>(rpm),
                      static_cast<unsigned>(load));
        assert_close(actual, expected, tol, label);
    }
}

void test_filter_moving_average(const json &ref, double tol) {
    const auto &spec = ref["stateful_filters"]["filter_moving_average"];
    SCE::Generated::FilterMovingAverage::FilterMovingAverage filter{};
    int i = 0;
    for (const auto &step : spec["sequence"]) {
        auto input = step["input"].get<double>();
        auto expected = step["expected"].get<double>();
        double actual = filter.update(input);
        char label[128];
        std::snprintf(label, sizeof(label),
                      "filter_moving_average step %d input=%.17g",
                      i, input);
        assert_close(actual, expected, tol, label);
        ++i;
    }
}

void test_filter_debounce(const json &ref) {
    const auto &spec = ref["stateful_filters"]["filter_debounce"];
    SCE::Generated::FilterDebounce::FilterDebounce filter{};
    int i = 0;
    for (const auto &step : spec["sequence"]) {
        auto input = step["input"].get<bool>();
        auto expected = step["expected"].get<bool>();
        bool actual = filter.update(input);
        CHECK(actual == expected,
              "filter_debounce step %d input=%d: actual=%d expected=%d",
              i,
              static_cast<int>(input),
              static_cast<int>(actual),
              static_cast<int>(expected));
        ++i;
    }
}

const char *coolant_tag_name(
    SCE::Generated::ObserverCoolant::ForgeDomain::Tag tag) {
    switch (tag) {
    case SCE::Generated::ObserverCoolant::ForgeDomain::EMIT_WARNING:
        return "EMIT_WARNING";
    case SCE::Generated::ObserverCoolant::ForgeDomain::CLEAR_WARNING:
        return "CLEAR_WARNING";
    case SCE::Generated::ObserverCoolant::ForgeDomain::EMERGENCY_SHUTDOWN:
        return "EMERGENCY_SHUTDOWN";
    }
    return "unknown";
}

void test_observer_coolant(const json &ref) {
    const auto &spec = ref["observers"]["observer_coolant"];
    SCE::Generated::ObserverCoolant::ObserverCoolant observer{};
    int i = 0;
    for (const auto &step : spec["sequence"]) {
        auto input = step["input"].get<double>();
        std::vector<std::string> expected_events;
        for (const auto &e : step["expected_events"]) {
            expected_events.push_back(e.get<std::string>());
        }
        auto queue = observer.update(input);
        std::vector<std::string> actual_events;
        for (std::size_t j = 0; j < queue.size(); ++j) {
            actual_events.emplace_back(coolant_tag_name(queue[j].tag));
        }
        if (actual_events != expected_events) {
            std::printf(
                "FAIL: observer_coolant step %d input=%.17g: got [",
                i, input);
            for (std::size_t j = 0; j < actual_events.size(); ++j) {
                if (j) std::printf(", ");
                std::printf("%s", actual_events[j].c_str());
            }
            std::printf("] expected [");
            for (std::size_t j = 0; j < expected_events.size(); ++j) {
                if (j) std::printf(", ");
                std::printf("%s", expected_events[j].c_str());
            }
            std::printf("]\n");
            ++g_failures;
        }
        ++i;
    }
}

void test_transform_temperature(const json &ref, double tol) {
    const auto &spec = ref["pure_functions"]["transform_temperature"];
    for (const auto &c : spec["cases"]) {
        auto raw = c["args"][0].get<std::uint16_t>();
        auto expected = c["expected"].get<double>();
        double actual =
            SCE::Generated::TransformTemperature::computeTemperature(raw);
        char label[128];
        std::snprintf(label, sizeof(label), "transform_temperature(%u)",
                      static_cast<unsigned>(raw));
        assert_close(actual, expected, tol, label);
    }
}

void test_transform_bitwise(const json &ref) {
    const auto &spec = ref["pure_functions"]["transform_bitwise"];
    for (const auto &c : spec["cases"]) {
        auto byte = c["args"][0].get<std::uint8_t>();
        auto expected_high = c["expected"]["high_nibble"].get<std::uint8_t>();
        auto expected_low = c["expected"]["low_nibble"].get<std::uint8_t>();
        auto actual_high = SCE::Generated::TransformBitwise::computeHighNibble(byte);
        auto actual_low = SCE::Generated::TransformBitwise::computeLowNibble(byte);
        CHECK(actual_high == expected_high,
              "transform_bitwise(%u) high: actual=%u expected=%u",
              static_cast<unsigned>(byte),
              static_cast<unsigned>(actual_high),
              static_cast<unsigned>(expected_high));
        CHECK(actual_low == expected_low,
              "transform_bitwise(%u) low: actual=%u expected=%u",
              static_cast<unsigned>(byte),
              static_cast<unsigned>(actual_low),
              static_cast<unsigned>(expected_low));
    }
}

void test_transform_multi_output(const json &ref, double tol) {
    const auto &spec = ref["pure_functions"]["transform_multi_output"];
    for (const auto &c : spec["cases"]) {
        auto celsius = c["args"][0].get<double>();
        auto expected_f = c["expected"]["fahrenheit"].get<double>();
        auto expected_k = c["expected"]["kelvin"].get<double>();
        double actual_f =
            SCE::Generated::TransformMultiOutput::computeFahrenheit(celsius);
        double actual_k =
            SCE::Generated::TransformMultiOutput::computeKelvin(celsius);
        char label[160];
        std::snprintf(label, sizeof(label),
                      "transform_multi_output(%.17g) fahrenheit", celsius);
        assert_close(actual_f, expected_f, tol, label);
        std::snprintf(label, sizeof(label),
                      "transform_multi_output(%.17g) kelvin", celsius);
        assert_close(actual_k, expected_k, tol, label);
    }
}

void test_condition_threshold(const json &ref) {
    const auto &spec = ref["pure_functions"]["condition_threshold"];
    for (const auto &c : spec["cases"]) {
        auto coolant = c["args"][0].get<double>();
        auto oil = c["args"][1].get<double>();
        auto max_temp = c["args"][2].get<double>();
        auto expected = c["expected"].get<bool>();
        bool actual = SCE::Generated::ConditionThreshold::conditionThreshold(
            coolant, oil, max_temp);
        CHECK(actual == expected,
              "condition_threshold(%.17g, %.17g, %.17g): actual=%d expected=%d",
              coolant, oil, max_temp,
              static_cast<int>(actual), static_cast<int>(expected));
    }
}

void test_condition_range(const json &ref) {
    const auto &spec = ref["pure_functions"]["condition_range"];
    for (const auto &c : spec["cases"]) {
        auto rpm = c["args"][0].get<std::uint32_t>();
        auto min_rpm = c["args"][1].get<std::uint32_t>();
        auto max_rpm = c["args"][2].get<std::uint32_t>();
        auto expected = c["expected"].get<bool>();
        bool actual = SCE::Generated::ConditionRange::conditionRange(
            rpm, min_rpm, max_rpm);
        CHECK(actual == expected,
              "condition_range(%u, %u, %u): actual=%d expected=%d",
              static_cast<unsigned>(rpm),
              static_cast<unsigned>(min_rpm),
              static_cast<unsigned>(max_rpm),
              static_cast<int>(actual), static_cast<int>(expected));
    }
}

void test_condition_programming(const json &ref) {
    const auto &spec = ref["pure_functions"]["condition_programming"];
    for (const auto &c : spec["cases"]) {
        auto engine_stop = c["args"][0].get<bool>();
        auto ignition = c["args"][1].get<bool>();
        auto expected = c["expected"].get<bool>();
        bool actual = SCE::Generated::ConditionProgramming::conditionProgramming(
            engine_stop, ignition);
        CHECK(actual == expected,
              "condition_programming(%d, %d): actual=%d expected=%d",
              static_cast<int>(engine_stop),
              static_cast<int>(ignition),
              static_cast<int>(actual), static_cast<int>(expected));
    }
}

void test_procedure_linear(const json &ref) {
    const auto &spec = ref["pure_functions"]["procedure_linear"];
    for (const auto &c : spec["cases"]) {
        auto value = c["args"][0].get<std::int32_t>();
        auto expected_completed = c["expected"]["completed"].get<bool>();
        auto expected_final = c["expected"]["final_state"].get<std::string>();
        SCE::Generated::ProcedureLinear::ProcedureLinear sm;
        auto result = sm.execute(value);
        CHECK(result.completed == expected_completed,
              "procedure_linear(%d).completed: actual=%d expected=%d",
              value,
              static_cast<int>(result.completed),
              static_cast<int>(expected_completed));
        CHECK(result.final_state == expected_final,
              "procedure_linear(%d).final_state: actual=%s expected=%s",
              value, result.final_state.c_str(), expected_final.c_str());
    }
}

void test_procedure_diamond(const json &ref) {
    const auto &spec = ref["pure_functions"]["procedure_diamond"];
    for (const auto &c : spec["cases"]) {
        auto sensor = c["args"][0].get<std::uint16_t>();
        auto mode = c["args"][1].get<std::string>();
        auto expected_completed = c["expected"]["completed"].get<bool>();
        auto expected_final = c["expected"]["final_state"].get<std::string>();
        SCE::Generated::ProcedureDiamond::ProcedureDiamond sm;
        auto result = sm.execute(sensor, mode);
        CHECK(result.completed == expected_completed,
              "procedure_diamond(%u, %s).completed: actual=%d expected=%d",
              static_cast<unsigned>(sensor), mode.c_str(),
              static_cast<int>(result.completed),
              static_cast<int>(expected_completed));
        CHECK(result.final_state == expected_final,
              "procedure_diamond(%u, %s).final_state: actual=%s expected=%s",
              static_cast<unsigned>(sensor), mode.c_str(),
              result.final_state.c_str(), expected_final.c_str());
    }
}

void test_procedure_startup_check(const json &ref) {
    const auto &spec = ref["pure_functions"]["procedure_startup_check"];
    for (const auto &c : spec["cases"]) {
        auto voltage = c["args"][0].get<float>();
        auto temperature = c["args"][1].get<float>();
        auto expected_completed = c["expected"]["completed"].get<bool>();
        auto expected_final = c["expected"]["final_state"].get<std::string>();
        SCE::Generated::ProcedureStartupCheck::ProcedureStartupCheck sm;
        auto result = sm.execute(voltage, temperature);
        CHECK(result.completed == expected_completed,
              "procedure_startup_check(%g, %g).completed: actual=%d expected=%d",
              static_cast<double>(voltage), static_cast<double>(temperature),
              static_cast<int>(result.completed),
              static_cast<int>(expected_completed));
        CHECK(result.final_state == expected_final,
              "procedure_startup_check(%g, %g).final_state: actual=%s expected=%s",
              static_cast<double>(voltage), static_cast<double>(temperature),
              result.final_state.c_str(), expected_final.c_str());
    }
}

} // namespace

int main() {
    json ref = load_reference();
    double tol = ref["float_tolerance"].get<double>();

    test_interpolation_1d_linear(ref, tol);
    test_interpolation_2d_bilinear(ref, tol);
    test_filter_moving_average(ref, tol);
    test_filter_debounce(ref);
    test_observer_coolant(ref);
    test_transform_temperature(ref, tol);
    test_transform_bitwise(ref);
    test_transform_multi_output(ref, tol);
    test_condition_threshold(ref);
    test_condition_range(ref);
    test_condition_programming(ref);
    test_procedure_linear(ref);
    test_procedure_diamond(ref);
    test_procedure_startup_check(ref);

    if (g_failures > 0) {
        std::printf("FAILED: %d assertion(s)\n", g_failures);
        return 1;
    }
    std::printf("OK: all conformance tests passed\n");
    return 0;
}

File: sce-forge-runtime/kotlin/src/jvmTest/kotlin/NumericalConformanceTest.kt
// SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2025 newmassrael
//
// Cross-language numerical conformance harness (Kotlin/JVM half).
//
// The fixtures referenced below are generated at configure time by the
// `generateForgeFixtures` Gradle task, which invokes sce-codegen on the
// SCXML files in tests/forge/resources/ and writes the Kotlin output into
// build/generated/conformance/kotlin/. That directory is wired into the
// jvmTest source set, so the generated classes land in this test binary's
// classpath alongside this file. No committed Kotlin goldens are consumed;
// the single source of truth is the SCXML and the codegen.
//
// All Kotlin Forge templates emit `package com.sce.generated.<name>`, so
// this test imports each fixture from its named package. The test file
// itself is package-free only because keeping it in the default package
// avoids having to pick an arbitrary package name for a single test class.

import com.sce.forge.runtime.EventQueue
import com.sce.generated.condition_programming.conditionProgramming
import com.sce.generated.condition_range.conditionRange
import com.sce.generated.condition_threshold.conditionThreshold
import com.sce.generated.filter_debounce.FilterDebounce
import com.sce.generated.filter_moving_average.FilterMovingAverage
import com.sce.generated.interpolation_1d_linear.Interpolation1dLinear
import com.sce.generated.interpolation_2d_bilinear.Interpolation2dBilinear
import com.sce.generated.observer_coolant.ForgeDomainTag
import com.sce.generated.observer_coolant.ObserverCoolant
import com.sce.generated.procedure_diamond.ProcedureDiamond
import com.sce.generated.procedure_diamond.ProcedureResult as DiamondResult
import com.sce.generated.procedure_linear.ProcedureLinear
import com.sce.generated.procedure_linear.ProcedureResult as LinearResult
import com.sce.generated.procedure_startup_check.ProcedureResult as StartupResult
import com.sce.generated.procedure_startup_check.ProcedureStartupCheck
import com.sce.generated.transform_bitwise.computeHighNibble
import com.sce.generated.transform_bitwise.computeLowNibble
import com.sce.generated.transform_multi_output.computeFahrenheit
import com.sce.generated.transform_multi_output.computeKelvin
import com.sce.generated.transform_temperature.computeTemperature
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.double
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Paths
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class NumericalConformanceTest {
    private val ref: JsonObject
    private val tol: Double

    init {
        val repoRoot = System.getProperty("sce.repo.root")
            ?: error("sce.repo.root system property not set by Gradle")
        val refPath = Paths.get(repoRoot, "tests", "forge", "conformance", "numerical_reference.json")
        ref = Json.parseToJsonElement(refPath.toFile().readText()).jsonObject
        tol = ref["float_tolerance"]!!.jsonPrimitive.double
    }

    private fun assertClose(actual: Double, expected: Double, label: String) {
        val diff = abs(actual - expected)
        assertTrue(
            diff <= tol,
            "$label: actual=$actual expected=$expected diff=$diff tol=$tol",
        )
    }

    private fun pureCases(name: String) =
        ref["pure_functions"]!!.jsonObject[name]!!.jsonObject["cases"]!!.jsonArray

    @Test
    fun interpolation1dLinearMatchesReference() {
        for (case in pureCases("interpolation_1d_linear")) {
            val rpm = case.jsonObject["args"]!!.jsonArray[0].jsonPrimitive.int.toUShort()
            val expected = case.jsonObject["expected"]!!.jsonPrimitive.double
            val actual = Interpolation1dLinear.lookup(rpm)
            assertClose(actual, expected, "interpolation_1d_linear($rpm)")
        }
    }

    @Test
    fun interpolation2dBilinearMatchesReference() {
        for (case in pureCases("interpolation_2d_bilinear")) {
            val args = case.jsonObject["args"]!!.jsonArray
            val rpm = args[0].jsonPrimitive.int.toUShort()
            val load = args[1].jsonPrimitive.int.toUByte()
            val expected = case.jsonObject["expected"]!!.jsonPrimitive.double
            val actual = Interpolation2dBilinear.lookup(rpm, load)
            assertClose(actual, expected, "interpolation_2d_bilinear($rpm, $load)")
        }
    }

    @Test
    fun filterMovingAverageMatchesReference() {
        val spec = ref["stateful_filters"]!!.jsonObject["filter_moving_average"]!!.jsonObject
        val filter = FilterMovingAverage()
        for ((i, step) in spec["sequence"]!!.jsonArray.withIndex()) {
            val input = step.jsonObject["input"]!!.jsonPrimitive.double
            val expected = step.jsonObject["expected"]!!.jsonPrimitive.double
            val actual = filter.update(input)
            assertClose(actual, expected, "filter_moving_average step $i input=$input")
        }
    }

    @Test
    fun filterDebounceMatchesReference() {
        val spec = ref["stateful_filters"]!!.jsonObject["filter_debounce"]!!.jsonObject
        val filter = FilterDebounce()
        for ((i, step) in spec["sequence"]!!.jsonArray.withIndex()) {
            val input = step.jsonObject["input"]!!.jsonPrimitive.boolean
            val expected = step.jsonObject["expected"]!!.jsonPrimitive.boolean
            val actual = filter.update(input)
            assertEquals(expected, actual, "filter_debounce step $i input=$input")
        }
    }

    @Test
    fun observerCoolantMatchesReference() {
        val spec = ref["observers"]!!.jsonObject["observer_coolant"]!!.jsonObject
        val observer = ObserverCoolant()
        for ((i, step) in spec["sequence"]!!.jsonArray.withIndex()) {
            val input = step.jsonObject["input"]!!.jsonPrimitive.double
            val expectedEvents = step.jsonObject["expected_events"]!!.jsonArray.map {
                it.jsonPrimitive.content
            }
            val queue: EventQueue<ForgeDomainTag> = observer.update(input)
            val actualEvents = queue.asList().map { it.name }
            assertEquals(expectedEvents, actualEvents, "observer_coolant step $i input=$input")
        }
    }

    @Test
    fun transformTemperatureMatchesReference() {
        for (case in pureCases("transform_temperature")) {
            val raw = case.jsonObject["args"]!!.jsonArray[0].jsonPrimitive.int.toUShort()
            val expected = case.jsonObject["expected"]!!.jsonPrimitive.double
            val actual = computeTemperature(raw)
            assertClose(actual, expected, "transform_temperature($raw)")
        }
    }

    @Test
    fun transformBitwiseMatchesReference() {
        for (case in pureCases("transform_bitwise")) {
            val byte = case.jsonObject["args"]!!.jsonArray[0].jsonPrimitive.int.toUByte()
            val expected = case.jsonObject["expected"]!!.jsonObject
            val expectedHigh = expected["high_nibble"]!!.jsonPrimitive.int.toUByte()
            val expectedLow = expected["low_nibble"]!!.jsonPrimitive.int.toUByte()
            val actualHigh = computeHighNibble(byte)
            val actualLow = computeLowNibble(byte)
            assertEquals(expectedHigh, actualHigh, "transform_bitwise($byte) high")
            assertEquals(expectedLow, actualLow, "transform_bitwise($byte) low")
        }
    }

    @Test
    fun transformMultiOutputMatchesReference() {
        for (case in pureCases("transform_multi_output")) {
            val celsius = case.jsonObject["args"]!!.jsonArray[0].jsonPrimitive.double
            val expected = case.jsonObject["expected"]!!.jsonObject
            val expectedF = expected["fahrenheit"]!!.jsonPrimitive.double
            val expectedK = expected["kelvin"]!!.jsonPrimitive.double
            val actualF = computeFahrenheit(celsius)
            val actualK = computeKelvin(celsius)
            assertClose(actualF, expectedF, "transform_multi_output($celsius) fahrenheit")
            assertClose(actualK, expectedK, "transform_multi_output($celsius) kelvin")
        }
    }

    @Test
    fun conditionThresholdMatchesReference() {
        for (case in pureCases("condition_threshold")) {
            val args = case.jsonObject["args"]!!.jsonArray
            val coolant = args[0].jsonPrimitive.double
            val oil = args[1].jsonPrimitive.double
            val maxTemp = args[2].jsonPrimitive.double
            val expected = case.jsonObject["expected"]!!.jsonPrimitive.boolean
            val actual = conditionThreshold(coolant, oil, maxTemp)
            assertEquals(expected, actual, "condition_threshold($coolant, $oil, $maxTemp)")
        }
    }

    @Test
    fun conditionRangeMatchesReference() {
        for (case in pureCases("condition_range")) {
            val args = case.jsonObject["args"]!!.jsonArray
            val rpm = args[0].jsonPrimitive.int.toUInt()
            val minRpm = args[1].jsonPrimitive.int.toUInt()
            val maxRpm = args[2].jsonPrimitive.int.toUInt()
            val expected = case.jsonObject["expected"]!!.jsonPrimitive.boolean
            val actual = conditionRange(rpm, minRpm, maxRpm)
            assertEquals(expected, actual, "condition_range($rpm, $minRpm, $maxRpm)")
        }
    }

    @Test
    fun conditionProgrammingMatchesReference() {
        for (case in pureCases("condition_programming")) {
            val args = case.jsonObject["args"]!!.jsonArray
            val engineStop = args[0].jsonPrimitive.boolean
            val ignition = args[1].jsonPrimitive.boolean
            val expected = case.jsonObject["expected"]!!.jsonPrimitive.boolean
            val actual = conditionProgramming(engineStop, ignition)
            assertEquals(expected, actual, "condition_programming($engineStop, $ignition)")
        }
    }

    @Test
    fun procedureLinearMatchesReference() {
        for (case in pureCases("procedure_linear")) {
            val value = case.jsonObject["args"]!!.jsonArray[0].jsonPrimitive.int
            val expected = case.jsonObject["expected"]!!.jsonObject
            val expectedCompleted = expected["completed"]!!.jsonPrimitive.boolean
            val expectedFinal = expected["final_state"]!!.jsonPrimitive.content
            val result: LinearResult = ProcedureLinear.execute(value)
            assertEquals(expectedCompleted, result.completed, "procedure_linear($value).completed")
            assertEquals(expectedFinal, result.finalState, "procedure_linear($value).final_state")
        }
    }

    @Test
    fun procedureDiamondMatchesReference() {
        for (case in pureCases("procedure_diamond")) {
            val args = case.jsonObject["args"]!!.jsonArray
            val sensor = args[0].jsonPrimitive.int.toUShort()
            val mode = args[1].jsonPrimitive.content
            val expected = case.jsonObject["expected"]!!.jsonObject
            val expectedCompleted = expected["completed"]!!.jsonPrimitive.boolean
            val expectedFinal = expected["final_state"]!!.jsonPrimitive.content
            val result: DiamondResult = ProcedureDiamond.execute(sensor, mode)
            assertEquals(expectedCompleted, result.completed, "procedure_diamond($sensor, $mode).completed")
            assertEquals(expectedFinal, result.finalState, "procedure_diamond($sensor, $mode).final_state")
        }
    }

    @Test
    fun procedureStartupCheckMatchesReference() {
        for (case in pureCases("procedure_startup_check")) {
            val args = case.jsonObject["args"]!!.jsonArray
            val voltage = args[0].jsonPrimitive.double.toFloat()
            val temperature = args[1].jsonPrimitive.double.toFloat()
            val expected = case.jsonObject["expected"]!!.jsonObject
            val expectedCompleted = expected["completed"]!!.jsonPrimitive.boolean
            val expectedFinal = expected["final_state"]!!.jsonPrimitive.content
            val result: StartupResult = ProcedureStartupCheck.execute(voltage, temperature)
            assertEquals(expectedCompleted, result.completed, "procedure_startup_check($voltage, $temperature).completed")
            assertEquals(expectedFinal, result.finalState, "procedure_startup_check($voltage, $temperature).final_state")
        }
    }
}
